prow/cmd/deck/main.go

```diff
@@ -139,6 +141,8 @@ func gatherOptions(fs *flag.FlagSet, args ...string) options {
 	fs.StringVar(&o.templateFilesLocation, "template-files-location", "/template", "Path to the template files")
 	fs.StringVar(&o.gcsCredentialsFile, "gcs-credentials-file", "", "Path to the GCS credentials file")
 	fs.BoolVar(&o.rerunCreatesJob, "rerun-creates-job", false, "Change the re-run option in Deck to actually create the job. **WARNING:** Only use this with non-public deck instances, otherwise strangers can DOS your Prow instance")
+	// csrfTokenFile, if specified, must point to a file containing a 32 byte token that will be used to protect against CSRF in POST requests
+	fs.StringVar(&o.csrfTokenFile, "csrf-token", "", "Path to the file containing the CSRF token.")
```

Comment:
Ideally we could reuse `--cookie-secret` for this - it serves the same purpose. Unfortunately, we didn't require 32-byte token (and it looks like on our deployment it's currently 64 bytes). We might be able to just truncate and complain if it's less than 32 bytes, but that presents a backwards compatibility concern. As far as I can tell, we never previously expressed guidance on how long `--cookie-secret` should be (or documented it at all?), which is itself bad.

Reply:
We're eventually going to need to announce that we'll start requiring a CSRF token anyway. We could instead announce that `--cookie-secret` must be at least 32 bytes (or some more specific length) and truncate it. What do you think?

Reply:
I think we should use `--cookie-secret` and do the following:

- if the value is more than 32 bytes, emit a warning that the value should be exactly 32 bytes and suggest truncating the existing value, then continue
- if the value is less than 32 bytes and `--rerun-creates-job` is _not_ specified but `--oauth-url` _is_, emit a warning that it should be exactly 32 bytes and continue
- if the value is less than 32 bytes and `--rerun-creates-job` _is_ specified, emit an error and refuse to start
- if neither `--rerun-creates-job` nor `--oauth-url` are specified, but `--cookie-secret` _is_, and `--cookie-secret` is not a 32-byte value, emit a warning and continue
- if none of `--rerun-creates-job`, `--oauth-url`, nor `--cookie-secret` are specified, do nothing and proceed silently

Note that for some reason, `--cookie-secret` is base64-encoded (and then k8s internally base64-encodes it again...), so all of the above is after decoding. Unfortunately, changing this would not be backwards compatible, so let's not.

Reply:
Done

Reply:
Are we sure that the entropy here is evenly distributed? If not then hashing the token might be preferable to truncating it.

Reply:
No way of knowing since AFAICT we never documented how to generate this (in our case, probably yes, but hard to tell for anyone else). Hashing is probably a good idea.

Reply:
Is there a specific library I should use for this?

Reply:
I'd suggest the built-in [`crypto/sha256`](https://golang.org/pkg/crypto/sha256/), which will give you the 32 bytes you want.